ocaml/typing/ctype.mli

```diff
@@ -486,6 +494,12 @@ val type_sort :
   why:Layouts.Layout.concrete_layout_reason ->
   Env.t -> type_expr -> (sort, Layout.Violation.t) result
 
+(* Same as [type_sort], but only safe to call on types known to be a sort.
+   For example, if the type is used as an argument in a function type that
+   has already been translated. *)
+val type_sort_exn :
```

Comment:
I don't think this should be particularly rare, because our choice not to store representations is quite intentional.  And I think it's usually cheap to recalculate these sorts, so I'm not particularly worried about that fact.

The obvious example is functions.  We've decided not to store sort or layout information in function types.  So when we go to check an application, and we need the sorts, we have to get recalculate them.  We do, on the other hand, store information in the typed tree, but that doesn't help us much because the thing being applied may not actually be a `Texp_function` - it could be any kind of term, all we know is that it has a function type, and there aren't any sorts there.  I'd expect this pattern to occur for basically all elimination forms.